src/main/java/jenkins/scm/api/SCMFileSystem.java

```diff
@@ -483,6 +552,25 @@ public final boolean supports(SCMSourceDescriptor descriptor) {
          */
         protected abstract boolean supportsDescriptor(SCMSourceDescriptor descriptor);
 
+        /**
+         * Given a {@link SCM} this should try to build a corresponding {@link SCMFileSystem} instance that
+         * reflects the content at the specified {@link SCMRevision}. If the {@link SCM} is supported but not
+         * for a fixed revision, best effort is acceptable as the most capable {@link SCMFileSystem} will be returned
+         * to the caller.
+         *
+         * @param build the build of the {@link SCM}
+         * @param scm the {@link SCM}.
+         * @param rev the specified {@link SCMRevision}.
+         * @return the corresponding {@link SCMFileSystem} or {@code null} if this builder cannot create a {@link
+         * SCMFileSystem} for the specified {@link SCM}.
+         * @throws IOException          if the attempt to create a {@link SCMFileSystem} failed due to an IO error
+         *                              (such as the remote system being unavailable)
+         * @throws InterruptedException if the attempt to create a {@link SCMFileSystem} was interrupted.
+         */
+        @CheckForNull
+        public abstract SCMFileSystem build(@NonNull Run build, @NonNull SCM scm, @CheckForNull SCMRevision rev)
```

Comment:
(In a perfect world where you could simultaneously release updates to all plugins that extend this class, and force all users to upgrade all of those plugins at the same time, this would not be a problem, but in reality, there may be some implementations of `SCMFileSystem.Builder` that are never updated to implement the new method, and users can choose to update some plugins but not others, so we need to preserve binary compatibility.)

Reply:
Please see: https://github.com/tgatinea/scm-api-plugin/pull/1